cpp_guidelines/guidelines.md

```diff
@@ -0,0 +1,496 @@
+  * `.clang-format`: `KeepEmptyLinesAtTheStartOfBlocks: false`
+  * `.clang-format`: `MaxEmptyLinesToKeep: 2`
+  * `.clang-format`: `SpaceAfterLogicalNot: false`
+  * `.clang-format`: `SpaceBeforeSquareBrackets: false`
+  * `.clang-format`: `SpaceInEmptyBlock: false`
+  * `.clang-format`: `SpaceInEmptyParentheses: false`
+  * `.clang-format`: `SpacesBeforeTrailingComments: 1`
+  * `.clang-format`: `SpacesInAngles: false`
+  * `.clang-format`: `SpacesInConditionalStatement: false`
+  * `.clang-format`: `SpacesInContainerLiterals: false`
+  * `.clang-format`: `SpacesInCStyleCastParentheses: false`
+  * `.clang-format`: `SpacesInParentheses: false`
+  * `.clang-format`: `SpacesInSquareBrackets: false`
+
+**Example:**
+```c++
+int                          // There is no reason for a line break
+foo();
+
+int bar( float x, float y ); // There is no reason for the spaces here
+float array[ 10 ];           // Or here
+baz< double >(...);          // Or here
+```
+
+### Use namespace end comments.
+
+**Reason:** Namespaces can span many lines. It is probably not obvious to
+which namespace a closing `}` belongs.
+
+**Enforced by:**
+  * `.clang-format`: `FixNamespaceComments: true`
+
+**Example:**
+```c++
+namespace bad_example
+{
+}
+
+namespace good_example
+{
+} // good_example
+```
+
+### Put nested namespaces on separate lines.
+
+If available the C++17 syntax is preferred: `namespace first::second`.
+
+**Reason**: Each statement belongs on its own line. This includes namespace
+declarations.
+
+**Enforced by:**
+  * `.clang-format`: `CompactNamespaces: false`
+
+**Example:**
+```c++
+// bad example
+namespace foo { namespace bar {
+}}
+
+// good example
+namespace foo
+{
+    namespace bar
+    {
+    }
+}
+
+// best example
+namespace foo::bar
+-{
+}
+```
+
+### Put constructor initializers and inheritance list members on separate lines.
+
+**Reason:** It is easier to read multiple entries one-by-one than having them
+arbitrarily spread across lines.
+
+**Exception**: All initializers / list members fit into one line.
+
+**Enforced by:**
+  * `.clang-format`: `ConstructorInitializerAllOnOneLineOrOnePerLine: true`
+
+**Example:**
+```c++
+class foo
+{
+    foo()
+        : aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa{val}
+        , bbbbbbbbbbbbb{another}
+        , ccccccccccccccccccccccccccc{value}
+    {}
+};
+```
+
+## Indentation and alignment.
+
+### Use the Allman indentation style.
+
+**Reason:** This gives the code a nice vertical structure and optically
+separates different scopes from another. The K&R indentation style
+unnecessarily increases code density, making it harder to follow its
+structure.
+
+**Enforced by:**
+  * `.clang-format`: `BreakBeforeBraces: Allman`
+
+**Example:**
+```c++
+// bad example
+if(foo) {
+    /* something */
+}
+
+// good example
+if(bar)
+{
+    /* something */
+}
+```
+
+### Use spaces, never tabs.
+
+**Reason:** Spaces are the same on all machines, tabs are not.
```

Comment:
It's good that tabs are not the same, so people can customize to what works best for them ;)

Reply:
I disagree. Customizing tab length individually would require the use of tabs in the code. Thus working together (in person) will lead to confusion.

Reply:
I do not understand your reasoning here: "Using tabs is bad because it requires tabs"? What @bernhardmgruber meant, was that the fact that the tab width is configurable locally is not a downside of indenting with tabs. It would be downside to aligning with tabs, but there are enough other reasons not to align (if aligning on should to this with spaces, i,e. `UseTab: AlignWithSpaces`).

Reply:
@jkelling Sorry for not communicating my point clearly. 

First of all, most codes in the CRP repo currently avoid the use of tabs (it is even forbidden in [picongpu](https://github.com/ComputationalRadiationPhysics/picongpu/blob/454a28efe6a6eb0fa1e0e97031ef2569a1ad313b/.travis.yml#L111-L114)), - thus tabs need to be added - most likely causing various issues if done automatically. That will lead to a lot of work. 

Secondly, it is not uncommon, at least for picongpu development, to work together in front of one screen. Using individual tab length will lead to confusion. Similarly, when copy-pasting code snippets to a browser (gitHub, gitLab,...), alignment with tabs is horrible. Thus complicating the discussion unnecessarily.

Reply:
* In this issue it appears, that the majority is against aligning. And I do agree that aligning with tabs is a bad idea because of the variability.
* Although I thought this style is foremost for future projects, not to be applied immediately for PIConGPU/Alpaka/.... conversion would not be hard:
  * If there is no alignment in the code base it would be simple to replace because there should be no clusters of spaces which are not indents, e.g. for `IndentWidth: 4`, do `s/    /\t/g`.
  * Non-indent space clusters exist it get a bit more complicated:
     ```bash
       sed -E -n ':start;s/^(\t*)  /\1\t/;tstart;p'
     ```
     (i.e. make sure to only replace space clusters at the beginnings of lines)
* Changing tab with does not change the structure of code, if two people who have different preferences use one screen, at least one of the would experience an indent width they are note used to, but they still see the same structure. One could have the same problem when looking at code from a different (not CRP) project anyway.
* I do not experience problems pasting code with tabs into github:
```c++
template<class T>
class Class
	: Base<T>
{
	int m_a = 0;

	public:

	Class(int a)
		: m_a(a)
	{
		assert(m_a < 3);
	}
};
```
You cannot type tabs in github, but any code formatting in these simple forms is tedious. It is easier to open an editor, type there and copy&paste either way. Apparently gitlab (or my browser?) has a tabwidth of eight.

Reply:
> Although I thought this style is foremost for future projects, not to be applied immediately for PIConGPU/Alpaka/.... conversion would not be hard: 

@jkelling the stylefile will be applied to alpaka, PIConGPU and other projects as soon as it is finished. see https://github.com/ComputationalRadiationPhysics/contributing/pull/39#issuecomment-694956183

Reply:
I think these discussions convinced me tabs are not evil. Using them for alignment is indeed a non-issue if (and righfully so imo anyways) alignment is not done.

Reply:
Let us just have a final vote on `UseTab` too, then.

**VOTE**
current: Never | AlignWithSpaces |
--------|-----------|
 0      |   1 |

* We found that some common arguments against tabs do not appear to apply for us:
  * All alignment options have been voted `false` based on other reasons.
  * The format file already contains `IndentWidth = TabWidth = 4`, with all other indents being the same width too. (Tangent: I do not understand why anyone would want `IndentWidth != TabWidth`, which weirdly is the default in vim (`sw = 4, ts = 8`), can someone enlighten me on this?)
  * Though tabs do not work in web forms, writing properly formatted code while having to input four spaces for each indent is probably to much of a pain anyway for anyone to actually do it and to fall back to an editor and copy&paste.
* Arguments for tabs over spaces, in order of my priority:
  1. Tabs are convenient even in the dumbest editors (press one time for indent, not four; same for unindent).
  2. Tabs are flexible: can set to less width if I happen to work on a small screen.
  3. Tabs discourage alignment.
  4. Tabs were made as field separators and thus are the natural choice for indentation (Note: not for alignment; to align with tabs one would have to always to one tab and then set the tab width so that all column contents are aligned, never use multiple tabs to get to some column)
* I see only one potential problem with tabs:
  1. Jupyter does not support inputting tabs: It can display them fine, but the tab key always indents with four spaces and there is no option to change this. Not sure if or when this issue will addressed by the developers. This may make it preferable for python programs to use spaces.
* Going for `AlignWithSpaces` even though there is no alignment in case any of the alignment options were to change later on.

Reply:
Never | AlignWithSpaces |
---------|------------------------|
1 | 1

Reply:
Never | AlignWithSpaces
-- | --
1 | 2

I played a bit in the LLAMA code base and with the rest of the settings we voted on. `AlignWithSpaces` seems reasonable to me.

Reply:
Never | AlignWithSpaces
-- | --
2 | 2

Reply:
Never | AlignWithSpaces
-- | --
3 | 2

Reply:
Voting closed, the winner is `Never`.